pkg/domain/node.go

```diff
@@ -245,30 +303,133 @@ func (n *Node) GrantVote(peerId string) {
 }
 
 func (n *Node) ApplyLogs() {
+	if n.exit {
+		return
+	}
+
+	hasApply := false
 	commitIndex := n.commitIndex
 	lastApplied := n.lastApplied
 
 	for lastApplied < commitIndex {
+		hasApply = true
 		// if lastApplied == 0 {
 		// 	c.initFsm()
 		// }
+
 		// Increment last applied first
 		// because lastApplied = 0 is not a valid logEntry
 		lastApplied += 1
 		if log, err := n.Log(lastApplied); err == nil {
-			if log.Type != LogCommand {
+			switch log.Type {
+			case LogCommand:
+				// do command
+				// res := c.evalFsm(log.Value, "COMMAND")
+				// if log.C != nil {
+				// 	log.C <- res
+				// }
+			case LogConfiguration:
+
+				var config ConfigurationUpdate
+				if err := json.Unmarshal(log.Data, &config); err == nil {
+					if res := n.configurationUpdate(config); log.C != nil {
+						log.C <- res
+					}
+				}
+			case LogNoop:
+				continue
+			default:
 				continue
 			}
-			// res := c.evalFsm(log.Value, "COMMAND")
-			// if log.C != nil {
-			// 	log.C <- res
-			// }
 		}
 	}
 
 	// Swap only once
-	if lastApplied > n.lastApplied {
+	if hasApply {
 		newState := n.withLastApplied(lastApplied)
 		n.swapState(&newState)
 	}
 }
+
+func (n *Node) ExecuteCommand(cmd ExecuteInput) chan ExecuteOutput {
+	result := make(chan ExecuteOutput, 1)
+	newEntry := LogEntry{
+		Index: uint64(n.lastLogIndex()),
+		Term:  n.currentTerm,
+		Type:  cmd.Type,
+		Data:  cmd.Data,
+		C:     result,
+	}
+
+	go n.dispatch(newEntry)
+
+	return result
+}
+
+func (n *Node) dispatch(entry LogEntry) {
+	n.AppendLogs(n.lastLogIndex(), entry)
+	n.synchronizeLogs()
+}
+
+// func (c ClusterNode) ExecuteQuery(query string) chan domain.ExecuteOutput {
+// 	result := make(chan domain.ExecuteOutput, 1)
+// 	go (func() { result <- c.evalFsm(query, "QUERY") })()
+// 	return result
+// }
+
+var (
+	errPeerAlreadyExists = errors.New("peer already exists")
+	errPeerDoesNotExist  = errors.New("peer does not exist")
+)
+
+func (n *Node) configurationUpdate(config ConfigurationUpdate) (res ExecuteOutput) {
```

Comment:
Cluster configuration use the log replication mechanism to ensure that the changes are duplicated on each node. In my opinion it is nice to reuse the log replication mechanism for managing the cluster state and not only duplicating FSM commands.